---
title: "GLM"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(vegan)
library(data.table)
library(dplyr)
library(reshape)
library(reshape2) #melt is masked from data.table
library(tibble)
library(ggplot2)
library(ape)
library(MCMCglmm)
library(picante)
library(tidyr)
library(geiger)
#library(modelsummary)
#library(gt)
#library(remotes)
```

##MCMC Generalized Linear Mixed-Effects Models  
Datasets: Species Cover by year/season, P/A

Classes of Analyses: 
1) Individual species within plots - analyze at species level w/ species' fates as a function of their phylogenetic position and trait characteristics. Aims to explain those aggregate effects as a function of what characteristics of species make them well-adapted or maladapted to environment of our field experiment.
2) Plot-level - See script 00.calculatePDmetrics:  using dFD and dPD metrics with taxonomic diversity change (SR) as a function of plot PD and FD. This tests of whether maintenance of diversity differs as a function of PD or FD or both.

#Phylogeny 
```{r}
phy <- read.tree(file="../DATA/phylogeny.analyzed.2016-01-05b.tre")
phy$tip.label[which(phy$tip.label == "Symphyotrichum_novae-angliae")] <-
  "Symphyotrichum_novae.angliae"
phy.tre <- drop.tip(phy, which(!phy$tip.label %in% names(subset(df.pa.plug.glm, select = Agalinis_tenuifolia:Zizia_aurea)))) #drop tips not in names(as.planted]
##Add spSite Phylogeny for GLM
#knittr()/source makeSpSite.R for t2 phylogeny
#source(file="./04a.MakeSiteSpTree.Rmd")
t2$node.label <- paste("node", 1:2147, sep="")
#phy.tre.ul <- chronopl(phy.tre, 0, age.min = 1, age.max = NULL)
phy.tre.ul <- chronopl(t2, lambda=1, age.min = 1, age.max = NULL)
#inv.phylo<-inverseA(phy.tre.ul,nodes="ALL", scale=TRUE)
inv.phylo <-inverseA(phy.tre.ul, nodes="ALL", scale=TRUE)
```
# TRAITS
```{r}
##################load MICE traits:
MICEtraits <- read.csv('../DATA/ImputedMiceTraits.2016-01-06.csv', as.is = T, row.names = 1)
setDT(MICEtraits, keep.rownames = "Species")[]
MICEtraits.drp <- subset(MICEtraits, select = -c(lifeCycleCoded, floweringStartcoded, wis_IL, habitClean, genome_Use) ) #drop
MICEtraits.drp$Species <- gsub("Symphyotrichum_novae-angliae", "Symphyotrichum_novae.angliae" , MICEtraits.drp$Species)
##Filter to only species in 2017 and 2018 cover (dropping spring = 105 species vs as.planted has 127 species
MICEtraits.drp <- MICEtraits.drp %>% filter (Species %in% names(subset(df.pa.plug.glm, select = Agalinis_tenuifolia:Zizia_aurea)))
MICEtraits.drp$Species -> rownames(MICEtraits.drp)
#MICEtraits.drp <- MICEtraits.drp[,-1]
#MICEtraits.drp2 <- subset(MICEtraits.drp, select= c(LDMC, SLA, SDMC, vegetativeHeight, LNC, seedMass)) #select 5 traits:
#match rownames(MICEtraits.drp2) to datCov.PA.long2$Species and fill in col values of MICEtraits.drp2:
#rownames(MICEtraits.drp2) -> MICEtraits.drp2$Species
#rownames(MICEtraits.drp) -> MICEtraits.drp$Species
MICEtraits.drp2 <- MICEtraits.drp

#################### load FULL traits matrix
FULLtraits <- read.csv('../DATA/traits.mat.updated.2019-08-12.csv', as.is = T, row.names = 1)
setDT(FULLtraits, keep.rownames = "Species")[]
FULLtraits$Species <- gsub("Symphyotrichum_novae-angliae", "Symphyotrichum_novae.angliae" , FULLtraits$Species)
```

##Data
```{r load and format data}
#df.pa.plug from 00c.phyD.SR.boxplots- changed to df.PA
block <- read.delim("../DATA/plots.blocking.v4.2016-08-03.tsv")
block <- subset(block, select=c(PLOT, block_BS.mod)).
df.cover.na.glm <- merge(df.cover.na, block, by.x="plot", by.y="PLOT")
df.cover.na.glm <- subset(df.cover.na.glm, select=-c(Code))
#df.cover.na.glm.long <- reshape2::melt(df.cover.na.glm,  id.vars = c("year", "plot", "type", "new.rep", "phyD", "fd", "block_BS.mod"), variable.name = "Species") 
df.cover.na.glm.long <- reshape2::melt(df.cover.na.glm,  id.vars = c("year", "plot", "type", "block_BS.mod"), variable.name = "Species") 
df.cover.na.glm.long2 <- df.cover.na.glm.long %>% drop_na() #drop NA rows
df.cover.na.glm.long.trait <- merge(df.cover.na.glm.long2, MICEtraits.drp2, by='Species')
df.cover.na.glm.long.trait <- df.cover.na.glm.long.trait[order(df.cover.na.glm.long.trait$plot),]
 #datCov.PA$plot -> rownames(datCov.PA2) #duplicate rownames not allowed.  
paste(df.cover.na.glm.long.trait$plot, df.cover.na.glm.long.trait$Species,sep="-") ->  df.cover.na.glm.long.trait$Species.new #Add phylogeny 

split(df.cover.na.glm.long.trait, df.cover.na.glm.long.trait$type) -> df.cover.na.glm.long.trait

as.numeric(df.cover.na.glm.long.trait$Plug$value) -> df.cover.na.glm.long.trait$Plug$value
###########ADD PHYLO
phy.tre.ul.p <- drop.tip(phy.tre.ul, which(!phy.tre.ul$tip.label %in% df.cover.na.glm.long.trait$Plug$Species.new))
inv.phylop <-inverseA(phy.tre.ul.p, nodes="ALL", scale=TRUE)
#
##########Runs models########################################
Covnull <- MCMCglmm(value ~ 1, data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=200, burnin=10000)
summary(Covnull) -> SCovnull

model.Cov.null <- MCMCglmm(value ~ 1, random=~plot, data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=200, burnin=10000)
summary(model.Cov.null) -> Smodel.Cov.null

model.Cov.null2 <- MCMCglmm(value ~ 1, random=~plot + block_BS.mod, data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=300, burnin=10000)
summary(model.Cov.null2) -> Smodel.Cov.null2

model.Cov.phy.null <- MCMCglmm(value ~ 1, random=~plot + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=df.cover.na.glm.long.trait$Plug,  nitt = 200000, thin=400, burnin=15000)
summary(model.Cov.phy.null) -> Smodel.Cov.phy.null

#model.Cov.phy.nullb <- MCMCglmm(value ~ 1, random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=200, burnin=10000)

model.Cov.phy.simple <- MCMCglmm(value ~ year, random=~plot + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=200, burnin=10000)        
summary(model.Cov.phy.simple) -> Smodel.Cov.phy.simple

model.Cov.phy.simpleb <- MCMCglmm(value ~ year, random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=df.cover.na.glm.long.trait$Plug,  nitt = 100000, thin=200, burnin=10000) 

#Rescale the fixed effects to mean 0 and unit variance: year?
Scaled.df.cover.na.glm.long.trait <- df.cover.na.glm.long.trait
cols <- c("seedMass",  "LDMC",   "SLA",  "LNC",   "LCC",  "LPC",  "SDMC", "circularity", "vegetativeHeight", "leafLength", "leafThickness", "leafWidth", "petioleLength")

Scaled.df.cover.na.glm.long.trait$Plug <- Scaled.df.cover.na.glm.long.trait$Plug %>%
  mutate_if(names(.) %in% cols, funs(c(scale(.))))
```
#Run more complex models for cover data (diagnostics below)
##traits, year interactions, etc.
```{r}
#make sure plot & year are factors
as.factor(Scaled.df.cover.na.glm.long.trait$Plug$plot) -> Scaled.df.cover.na.glm.long.trait$Plug$plot
as.numeric(Scaled.df.cover.na.glm.long.trait$Plug$value) -> Scaled.df.cover.na.glm.long.trait$Plug$value

############Now traits:################################################
model.traits <- MCMCglmm(value ~ seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth + Lifeform + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer, random=~plot, data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits) -> Smodel.traits

model.traits.year <- MCMCglmm(value ~ year + seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth+ Lifeform + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer, random=~plot + block_BS.mod, data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits.year) -> Smodel.traits.year

model.traits.phy <- MCMCglmm(value ~ seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth+ Lifeform + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer, random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits.phy) -> Smodel.traits.phy

model.traits.phy.year <- MCMCglmm(value ~ year + seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth+ Lifeform + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer, random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits.phy.year) -> Smodel.traits.phy.year

model.traits1 <- MCMCglmm(value ~ year + seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth + Lifeform + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer + (seedMass:year) + (LDMC:year) + (SLA:year) + (LNC:year) + (LCC:year) + (LPC:year) + (SDMC:year) + (vegetativeHeight:year) + (Lifeform:year) + (rhizomes:year) + (bulbs:year) + (corms:year) + (fibrous:year) + (leafLength:year) + (nFixer:year), random=~plot + block_BS.mod, data=Scaled.df.cover.na.glm.long.trait$Plug,  nitt = 200000, thin=200, burnin=20000)
summary(model.traits1) -> Smodel.traits1

model.traits.phy.int <- MCMCglmm(value ~ year + seedMass + LDMC + SLA + LNC + LCC + LPC + SDMC + circularity + vegetativeHeight + leafLength + leafThickness + leafWidth + photosyntheticPathway + rhizomes + bulbs + corms + fibrous + nFixer + (seedMass:year) + (LDMC:year) + (SLA:year) + (LNC:year) + (LCC:year) + (LPC:year) + (SDMC:year) + (vegetativeHeight:year) + (rhizomes:year) + (bulbs:year) + (corms:year) + (fibrous:year) + (leafLength:year) + (nFixer:year), random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits.phy.int) -> Smodel.traits.phy.int

model.traits.phy.intsimple <- MCMCglmm(value ~ (seedMass:year) + (LDMC:year) + (SLA:year) + (LNC:year) + (LCC:year) + (LPC:year) + (SDMC:year) + (vegetativeHeight:year) + (Lifeform:year) + (rhizomes:year) + (bulbs:year) + (corms:year) + (fibrous:year) + (leafLength:year) + (nFixer:year), random=~plot + block_BS.mod + Species.new, ginverse= list(Species.new=inv.phylop$Ainv), data=Scaled.df.cover.na.glm.long.trait$Plug, nitt = 200000, thin=200, burnin=20000)
summary(model.traits.phy.intsimple) -> Smodel.traits.phy.intsimple

```


#Optimal Model(s) diagnostics
```{r}
#pr = TRUE, To examine the random intercepts, one should add pr = TRUE (FALSE by default).
######select prior G=list(G1=list(V=1, nu=0.002))) #for phylogeny as random effect (http://www.jcsantosresearch.org/Class_2014_Spring_Comparative/pdf/week_12/Mar_24_2015_continuous_discrete_ver_4.pdf)
effectiveSize(model.PA$Sol)
#the models should have over-dispersed starting values. This will stop MCMCglmm finding heuristically good starting values.
#Approximate convergence is diagnosed when the upper CI limit is close to 1
gelman.diag(mcmc.list(model.PA$Sol, model.PA1$Sol)) #have to have same nitt and thinning
#autocorrelations
autocorr.plot(model.PA$Sol)
autocorr.plot(model.PA$VCV)
xyplot(log(model.PA$VCV))
plot(model.phy.simpleb$VCV) #check random effects trace plot

#robustness to different priors
  ## Parameter-expanded prior should be uninformative for variances and covariance

#parameter expanded priors for the random effects by including prior = prior1. Each random effect is represented by a G, and the residual is represented by R. The parameter expansion refers to the fact that we have included a prior mean (alpha.mu) and (co)variance matrix (alpha.V) as well as V and nu. For now, alpha.V is going to be 1000
a <- 1000
prior1 <- list(R = list(V = diag(1), nu = 0.002),
               G = list(G1 = list(V = diag(1), nu = 1, alpha.mu = 0, alpha.V = diag(1)*a),
                        G1 = list(V = diag(1), nu = 1, alpha.mu = 0, alpha.V = diag(1)*a),
                        G1 = list(V = diag(1), nu = 1, alpha.mu = 0, alpha.V = diag(1)*a)))

#randomprior <- MCMCglmm( ,data=, prior = prior1, nitt = 60000) #If effective sample sizes are much bigger; This is a good sign.
#plot(randomprior$VCV)
```


```{r}
#Making summary tables and plot effect size
SSS <- vector('list', 10)
glmTable <- vector('list', 10)

SSSnames <- c("Smodel.Cov.null", "Smodel.Cov.null2", "Smodel.Cov.phy.null","Smodel.Cov.phy.simple", "Smodel.traits", "Smodel.traits1", "Smodel.traits.year", "Smodel.traits.phy.year", "Smodel.traits.phy.intsimple", "Smodel.traits.phy.int")

#SSS[[1]] <- SCovnull #add manually due to no Cgov
SSS[[1]] <- Smodel.Cov.null 
SSS[[2]] <- Smodel.Cov.null2
SSS[[3]] <- Smodel.Cov.phy.null
SSS[[4]] <- Smodel.Cov.phy.simple
SSS[[5]] <- Smodel.traits
SSS[[6]] <- Smodel.traits1
SSS[[7]] <- Smodel.traits.year
SSS[[8]] <- Smodel.traits.phy.year
SSS[[9]] <- Smodel.traits.phy.intsimple
SSS[[10]] <- Smodel.traits.phy.int

#preds <- c("year", 'seedMass', 'LDMC',"SLA", "LNC", "LCC", "LPC", "SDMC", "circularity", "vegetativeHeight", "leafLength","leafThickness", "leafWidth", "rhizomes", "bulbs","corms" ,"fibrous", "nFixer")

##Intsead use rownames(SSS[[9]]$solutions to fill in preds
#loop to create final preds based on match update (to get most robust preds list)
preds1 <- list(rownames(SSS[[9]]$solutions))
preds <- unlist(rownames(SSS[[10]]$solutions))
#unlist(list(preds1, preds2)) -> newlist
#unique(newlist) -> preds

gcovs <- c('block_BS.mod', 'plot', 'Species.new')
headers <- c('model', 'DIC', 'Residuals', gcovs, preds)

glmmSummary <- matrix(NA, 0, length(headers), dimnames = list(NULL, headers))
roundTo = 4

for(i in 1:length(SSS)) {
      mTemp <- SSS[[i]] 
      glmmSummary <- rbind(glmmSummary,
                     c(SSSnames[[i]], #name of dataset in same order
                       DIC = round(mTemp$DIC, roundTo),
                       Residuals = mTemp$Rcovariances[1:3] %>%
                         round(roundTo) %>%
                         format(digits = roundTo) %>%
                         (function(x) paste(x[1], '(', x[2], ',',
                                            x[3], ')', sep = '')),
                
                       as.data.frame(mTemp$Gcovariances)[gcovs, 1:3] %>%
                         round(roundTo) %>%
                         format(digits = roundTo) %>%
                         apply(., 1, function(x) paste(x[1],
                                                       ' (', x[2],',',
                                                       x[3],')', sep = '')), 
                       
                        as.data.frame(mTemp$solutions)[preds, 1:3] %>% #how to paste based on matching name and predefined order?
                         round(roundTo) %>%
                         format(digits = roundTo) %>%
                         apply(., 1, function(x) paste(x[1],
                                                       ' (',x[2], ',',
                                                       x[3],')', sep = ''))
   ) #close c
 ) #close rbind
} #close coverType
 
as.data.frame(glmmSummary) ->glmmSummary
write.csv(glmmSummary, file="../OUT/glmmSummary_all.csv")
```

```{r}
SSS <- Smodel.traits.phy.int
for (i in 1:length(SSS)) {
  as.data.frame(SSS$solutions) -> fixed
  as.data.frame(SSS$Gcovariances) -> random
  as.data.frame(SSS$Rcovariances) -> Rvar
  SSS$DIC -> dic
headers <- c("modelName", "DIC", "predictors", colnames(SSS$solutions))
c(rownames(fixed), rownames(random), rownames(Rvar)) -> preds
c(rownames(fixed), rownames(Rvar)) -> preds.new

bind_rows(fixed, random, Rvar) -> dt #drops row and col names
bind_rows(fixed, Rvar) -> dt.new #drops row and col names

names(dt) <- c("post.mean", "CI_low", "CI_high", "eff.samp", "pMCMC") #fix names
names(dt.new) <- c("post.mean", "CI_low", "CI_high", "eff.samp", "pMCMC") #fix names
data.frame(Modelname=paste(SSSnames[i]), DIC = dic, Predictors=unlist(preds), dt) -> Temp
Temp -> glmTable
}

rbind(glmTable[[1]],glmTable[[2]] ) -> Table_glmSummary
write.csv(glmTable, file="../OUT/Smodel.traits.phy.int.csv")
```
 
```{r}
#######effect size plot ############## 
p <- ggplot(glmTableToPlot[2:26,], mapping = aes(x = reorder(Predictors, post.mean),
                                    y = post.mean, ymin = CI_low, ymax = CI_high))
effsam.plot <- p + geom_pointrange() + coord_flip() + labs(x="", y="post mean and CIs")
#ggsave(effsam.plot, file="../OUT/model.PA.phy_effectSize.png", width=80, units="mm")
#now plot multiple models, model each colored, to compare:

#color based on catageory:
years <- c("year2017", "year2018.05", "year2018.09", "year2019.06", "year2019.09")
roots <- c("rhizomes", "bulbs", "corms", "fibrous")
#random <- c("units", "Species.new", "block_BS.mod", "plot", "(Intercept)")
lifeform <- c("Lifeformhemicryptophytes", "Lifeformphanerophytes", "Lifeformgeophytes")
traits <- c("leafWidth", "leafThickness", "leafLength", "circularity", "SLA", "LNC", "LCC", "LDMC", "LPC", "vegetativeHeight", "seedMass", "SDMC")
other <- c("nFixer", "photosyntheticPathwayC4")

new.groups <- data.table(Predictors=c(years, roots, lifeform, traits, other), Group=c( 
  rep("year", length(years)), 
  rep("roots",length(roots)), 
  rep("lifeform", length(lifeform)), 
    rep("random", length(random)), 
  rep("traits", length(traits)), 
  rep("other", length(other)) ), key=c("Predictors", "Group"))  

new.glmTable <- merge(glmTable, new.groups, by.x="Predictors", by.y="Predictors") 

p2 <- ggplot(new.glmTable, mapping = aes(x = reorder(Predictors, post.mean),
                                    y = post.mean, ymin = CI_low, ymax = CI_high, color=Group))
effsam.plot2 <- p2 + geom_pointrange() + coord_flip() + labs(x="", y="post mean and CIs") + theme_bw() + 
  scale_colour_manual(values = pal_jco()(6), name="Predictors")
    
```